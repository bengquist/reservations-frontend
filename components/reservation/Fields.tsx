import * as React from "react";
import DatePicker from "react-datepicker";
import Section from "../ui/Section";
import styled from "styled-components";

type Values = {
  name: string;
  hotelName: string;
  arrivalDate: Date;
  departureDate: Date;
  errorMessage: string;
};

type Props = {
  values: Values;
  onInput: (value: any) => void;
};

const Fields: React.FunctionComponent<Props> = ({ onInput, values }) => {
  const { name, hotelName, arrivalDate, departureDate } = values;

  return (
    <>
      <Section
        label="Reservation Name:"
        placeholder="Blake"
        value={name}
        setValue={(value: string) => onInput({ name: value })}
      />
      <Section
        label="Hotel:"
        placeholder="Hampton Inn & Suites"
        value={hotelName}
        setValue={(value: string) => onInput({ hotelName: value })}
      />
      <DateContainer>
        <Section label="Arrival Date:" value={Number(arrivalDate)}>
          <DatePicker
            selected={arrivalDate}
            onChange={(value: Date) => onInput({ arrivalDate: value })}
          />
        </Section>
        <Section label="Departure Date:" value={Number(departureDate)}>
          <DatePicker
            selected={departureDate}
            onChange={(value: Date) => onInput({ departureDate: value })}
          />
        </Section>
      </DateContainer>
    </>
  );
};

export default Fields;

const DateContainer = styled.div`
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-gap: 2rem;

  @media (max-width: 500px) {
    grid-template-columns: 1fr;
  }
`;
