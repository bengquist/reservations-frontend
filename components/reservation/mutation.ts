import gql from "graphql-tag";

export const RESERVATION_MUTATION = gql`
  mutation RESERVATION_MUTATION(
    $name: String!
    $hotelName: String!
    $arrivalDate: String!
    $departureDate: String!
  ) {
    addReservation(
      name: $name
      hotelName: $hotelName
      arrivalDate: $arrivalDate
      departureDate: $departureDate
    ) {
      name
      hotelName
      arrivalDate
      departureDate
      id
    }
  }
`;
