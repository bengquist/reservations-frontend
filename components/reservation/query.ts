import gql from "graphql-tag";

export const RESERVATIONS_QUERY = gql`
  query RESERVATIONS_QUERY {
    reservations {
      name
      hotelName
      arrivalDate
      departureDate
      id
    }
  }
`;

export const RESERVATION_QUERY = gql`
  query RESERVATION_QUERY($id: String!) {
    reservation(id: $id) {
      name
      hotelName
      arrivalDate
      departureDate
      id
    }
  }
`;
