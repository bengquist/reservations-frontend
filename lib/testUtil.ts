import casual from "casual";

casual.seed(333);

const mockReservation = () => ({
  arrivalDate: "06/09/2019",
  departureDate: "06/19/2019",
  hotelName: "Hilton Garden Inn",
  id: "j1h23489ydsa9sf8ha",
  name: "Blake"
});

const mockReservationList = () => [
  {
    arrivalDate: "06/09/2019",
    departureDate: "06/19/2019",
    hotelName: "Hilton Garden Inn",
    id: casual.uuid,
    name: "Blake"
  },
  {
    arrivalDate: "06/02/2019",
    departureDate: "06/10/2019",
    hotelName: "Homewood Suites",
    id: casual.uuid,
    name: "Ashlynn"
  },
  {
    arrivalDate: "06/05/2019",
    departureDate: "06/14/2019",
    hotelName: "The Statler Dallas",
    id: casual.uuid,
    name: "Tammy"
  },
  {
    arrivalDate: "06/11/2019",
    departureDate: "06/16/2019",
    hotelName: "Hampton Inn & Suites",
    id: casual.uuid,
    name: "Jessica"
  }
];

export { mockReservation, mockReservationList, casual };
