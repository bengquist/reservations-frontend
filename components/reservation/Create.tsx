import * as React from "react";
import { Reservation, ReservationUpdate } from "./types";
import { RESERVATIONS_QUERY } from "./query";
import { Mutation } from "react-apollo";
import { RESERVATION_MUTATION } from "./mutation";
import { format } from "date-fns";
import { Message, Form, ButtonContainer, secondaryStyles } from "./styles";
import { preventDefault } from "../../lib/eventHelpers";
import Link from "next/link";
import Button from "../ui/Button";
import Fields from "./Fields";
import "react-datepicker/dist/react-datepicker-cssmodules.css";

type Data = {
  addReservation: Reservation[];
};

const Create = () => {
  const defaultValues = {
    name: "",
    hotelName: "",
    arrivalDate: new Date(),
    departureDate: new Date(),
    errorMessage: ""
  };

  const updateReservations: ReservationUpdate = (
    cache,
    { data: { addReservation } }
  ) => {
    const { reservations } = cache.readQuery({ query: RESERVATIONS_QUERY });

    cache.writeQuery({
      query: RESERVATIONS_QUERY,
      data: { reservations: [...reservations, addReservation] }
    });
  };

  const [inputValues, setInputValues] = React.useState(defaultValues);
  const { name, hotelName, arrivalDate, departureDate } = inputValues;

  return (
    <Mutation<Data>
      mutation={RESERVATION_MUTATION}
      variables={{
        name,
        hotelName,
        arrivalDate: format(arrivalDate, "MM/DD/YYYY"),
        departureDate: format(departureDate, "MM/DD/YYYY")
      }}
      update={updateReservations}
    >
      {(addReservation, { data }) => {
        const submitHandler = async () => {
          await addReservation();
          setInputValues(defaultValues);
        };

        const isSubmitted = data && data.addReservation;
        const message = isSubmitted && (
          <Message>Reservation successfully added!</Message>
        );

        return (
          <Form onSubmit={preventDefault(submitHandler)}>
            <Fields
              values={inputValues}
              onInput={(value: any) =>
                setInputValues(values => {
                  return { ...values, ...value };
                })
              }
            />
            {message}
            <ButtonContainer>
              <Link href="/">
                <Button type="button">
                  {isSubmitted ? "Go Home" : "Cancel"}
                </Button>
              </Link>
              <Button type="submit" css={secondaryStyles}>
                Submit
              </Button>
            </ButtonContainer>
          </Form>
        );
      }}
    </Mutation>
  );
};

export default Create;
