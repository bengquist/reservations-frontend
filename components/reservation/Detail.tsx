import gql from "graphql-tag";
import { Query } from "react-apollo";
import styled from "styled-components";
import { formatDate } from "../../lib/formatHelpers";

type Props = {
  id: string;
};

export const RESERVATION_QUERY = gql`
  query RESERVATION_QUERY($id: String!) {
    reservation(id: $id) {
      name
      hotelName
      arrivalDate
      departureDate
      id
    }
  }
`;

const Detail: React.FunctionComponent<Props> = ({ id }) => {
  return (
    <Query query={RESERVATION_QUERY} variables={{ id }}>
      {({ data, loading }) => {
        if (loading) return null;
        const {
          name,
          hotelName,
          arrivalDate,
          departureDate
        } = data.reservation;

        return (
          <Container>
            <h1>Reservation</h1>
            <div className="info">
              <p>
                <span>Name: </span>
                {name}
              </p>
              <p>
                <span>Hotel Name: </span> {hotelName}
              </p>
              <p>
                <span>Arrival Date: </span> {formatDate(arrivalDate)}
              </p>
              <p>
                <span>Departure Date: </span> {formatDate(departureDate)}
              </p>
            </div>
          </Container>
        );
      }}
    </Query>
  );
};

export default Detail;

const Container = styled.div`
  > h1 {
    background: ${props => props.theme.colors.gray};
    text-transform: uppercase;
    font-weight: 700;
    color: ${props => props.theme.colors.accent};
    text-align: center;
    padding: 1.5rem;
    box-shadow: ${props => props.theme.shadows.soft};
  }

  > .info {
    display: grid;
    grid-gap: 5px;
    padding: 2rem;

    > p > span {
      color: ${props => props.theme.colors.accent};
      font-weight: 700;
    }
  }
`;
