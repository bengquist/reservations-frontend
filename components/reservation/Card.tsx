import { Reservation } from "./types";
import styled, { css } from "styled-components";
import { hoverState, focusState } from "../styles/helpers";
import { formatDate } from "../../lib/formatHelpers";
import Modal from "../ui/Modal";
import Detail from "./Detail";
import { useState } from "react";

type Props = {
  reservation: Reservation;
};

const Card: React.FunctionComponent<Props> = ({ reservation }) => {
  const [showModal, setShowModal] = useState();

  return (
    <>
      <button onClick={() => setShowModal(true)} css={buttonStyle}>
        <Container>
          <span>{reservation.name}</span>
          <span>{reservation.hotelName}</span>
          <span className="date">{formatDate(reservation.arrivalDate)}</span>
          <span className="date">{formatDate(reservation.departureDate)}</span>
        </Container>
      </button>
      {showModal && (
        <Modal onClose={() => setShowModal(false)}>
          <Detail id={reservation.id} />
        </Modal>
      )}
    </>
  );
};

export default Card;

const buttonStyle = css`
  width: 100%;
  background: white;
  ${hoverState};
  ${focusState};
`;

const Container = styled.div`
  display: grid;
  grid-template-columns: 1.25fr 1.25fr 0.75fr 0.75fr;
  text-align: left;

  > span {
    padding: 1rem 0.5rem;
    border-right: 2px solid
      ${props => props.theme.colors && props.theme.colors.gray};
    :last-child {
      border-width: 1px 0 1px 1px;
    }
  }

  > .date {
    text-align: center;
  }

  @media (max-width: 500px) {
    font-size: 1rem;
  }
`;
