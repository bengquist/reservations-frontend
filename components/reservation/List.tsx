import Card from "./Card";
import { Reservation } from "./types";
import styled from "styled-components";

type Props = {
  reservations: Reservation[];
};

const List: React.FunctionComponent<Props> = ({ reservations }) => {
  const renderReservation = (reservation: Reservation) => {
    return <Card key={reservation.id} reservation={reservation} />;
  };

  return <CardsContainer>{reservations.map(renderReservation)}</CardsContainer>;
};

export default List;

const CardsContainer = styled.div`
  box-shadow: ${props => props.theme.shadows.soft};
`;
