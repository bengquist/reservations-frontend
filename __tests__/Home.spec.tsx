import * as React from "react";
import { mount, shallow } from "enzyme";
import List from "../components/reservation/List";
import { mockReservationList } from "../lib/testUtil";
import { RESERVATIONS_QUERY } from "../components/reservation/query";
import { MockedProvider } from "react-apollo/test-utils";

describe("Home", () => {
  it("renders and matches snapshot", () => {
    const wrapper = shallow(<List reservations={mockReservationList()} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("renders list with correct data", async () => {
    const mocks = [
      {
        request: {
          query: RESERVATIONS_QUERY
        },
        result: {
          data: {
            reservations: mockReservationList()
          }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <List reservations={mockReservationList()} />
      </MockedProvider>
    );

    expect(wrapper.find("span").length).toEqual(16);
  });
});
