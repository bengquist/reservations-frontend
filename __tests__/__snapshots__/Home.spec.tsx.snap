// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders and matches snapshot 1`] = `ShallowWrapper {}`;
